wdns/str_to_rdata_ubuf.c

```diff
@@ -934,22 +971,121 @@ _wdns_str_to_rdata_ubuf(ubuf *u, const char *str,
 		}
 
 		case rdf_svcparams: {
-			/* process ony key=val pair at a time */
-			char *buf, *tok, *endp = NULL;
+			char *start, *eol;
+			int prev_key = -1;
+
+			eol = (char *)str;
+			while (eol != NULL && *eol != '\0') {
+				eol++;
+			}
+
+			start = (char *)str;
+
+			while (start < eol) {
+				char *key, *val, *end, *dup;
+				uint16_t k;
+				bool quotes;
+
+				/* find out what key we're parsing */
+				if (*start == ' ') {
+					start++;
+					continue;
+				}
+
+				key = start;
+				val = start;
+
+				/*
+				 * The 'key' component of a parameter must end
+				 * with a:
+				 *  - '=' sign (preceeding a value);
+				 *  - blank space or eol (if that particular
+				 *    key doesn't require one).
+				 */
+				while (val != eol) {
+					if (*val == '=' || *val == ' ') {
+						break;
+					}
+					val++;
+				}
 
-			buf = strndup(str, strlen(str));
-			tok = strtok_r(buf, " \t\r\n", &endp);
+				dup = strndup(key, val - key);
+				k = _wdns_str_to_svcparamkey(dup);
+				free(dup);
+
+				/*
+				 * Fail if the key:
+				 *  - is invalid;
+				 *  - it's numeric value is smaller than the
+				 *    previous key (i.e they're not in
+				 *    ascending order);
+				 *  - doesn't require a value but has one;
+				 *  - requires a value but has none.
+				 */
+				if (k == spr_invalid) {
+					return (wdns_res_parse_error);
+				}
 
-			while (tok != NULL) {
-				res = str_to_svcparam(u, tok);
+				if (k <= prev_key) {
+					return (wdns_res_parse_error);
+				}
+
+				prev_key = k;
+
+				if (*val == ' ' || val == eol) {
+					if (k != spr_nd_alpn) {
+						return (wdns_res_success);
+					}
+
+					/* process spr_nd_alpn here */
+					res = str_to_svcparam(u, k, NULL);
+					if (res != wdns_res_success) {
+						return (res);
+					}
+
+					start = val;
+					continue;
+				}
+
+				if (*val != '=') {
+					return (wdns_res_parse_error);
+				}
+
+				/* now parse the value */
+				end = ++val;
+
+				if (*end == '"') {
+					quotes = true;
+					end++;
+				} else {
+					quotes = false;
+				}
+
+				/* values always end with a blank space */
+				while (end < eol) {
+					if (*end == '"' && quotes) {
+						/* end of double quotes */
+						end++;
+						break;
+					}
+					if (*end == ' ' && !quotes) {
+						/* end of param */
+						break;
+					}
+					end++;
+				}
+
+				dup = strndup(val, end - val);
+
+				res = str_to_svcparam(u, k, dup);
 				if (res != wdns_res_success) {
-					goto err;
+					return (res);
 				}
 
-				tok = strtok_r(NULL, " ", &endp);
+				free(dup);
```

Comment:
This `free()` needs to be moved earlier to avoid leaking if `res != wdns_res_success`.

Note that if `str_to_svcparam` could avoid using `strtok_r` (similar parsing code elsewhere in wdns uses `strpbrk()` to non-destructively scan for separators), it could accept a `const char *` and make `strdup()`/`free()` pairs unnecessary.

Reply:
I use `strdup()` after identifying the `value` portion of `key=value`, before calling `str_to_svcparam()`. That said, about half the uses of `strpbrk()` in wdns also use `strdup()` and I'm not sure that saving on that small alloc/free is worth it. I also looked at implementing the `strtok_r()` pac-man style loop with `strpbrk()` and it wasn't as clean to me.

Reply:
Fair, and agreed that `strtok_r` is much more convenient when it works (which it does here, after `strdup()`).

Leaving this comment open until `dup` is no longer leaked on `str_to_svcparam` failure.